main/server.go

```diff
@@ -0,0 +1,313 @@
+}
+
+// badRequest should be returned by a handler function to indicate a malformed request.
+type badRequest struct{ error }
+
+// notFound should be returned by a handler function to indicate that the requested data could not be found.
+type notFound struct{ error }
+
+// IsApiEnabled writes an HTTP 200 indicating that the TodoMvc API is enabled for this app.
+func IsApiEnabled(w http.ResponseWriter, r *http.Request) {
+	w.WriteHeader(http.StatusOK)
+}
+
+// errorHandler wraps a function returning an error by handling the error and returning a http.Handler.
+// If the error is of the one of the types defined above, it is handled as described for every type.
+// If the error is of another type, it is considered as an internal error and its message is logged.
+func errorHandler(f func(w http.ResponseWriter, r *http.Request, c context.Context) error) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		c := appengine.NewContext(r)
+		err := f(w, r, c)
+		if err == nil {
+			return
+		}
+		switch err.(type) {
+		case badRequest:
+			http.Error(w, err.Error(), http.StatusBadRequest)
+		case notFound:
+			http.Error(w, "todo not found", http.StatusNotFound)
+		default:
+			log.Errorf(c, "internal error %v", err)
+			http.Error(w, "Internal Server Error", http.StatusInternalServerError)
+		}
+	}
+}
+
+// writeJson marshals v into its JSON representation and writes that to w
+// along with the given status code. It also takes care to set the HTTP
+// Content-Type header. Any error encountered during marshalling is returned.
+func writeJson(w http.ResponseWriter, statusCode int, v interface{}) error {
+	w.Header().Set("Content-Type", "application/json; charset=utf-8")
+	w.WriteHeader(statusCode)
+	return json.NewEncoder(w).Encode(v)
+}
+
+// ListTodos handles GET requests on /todos.
+// It requires no parameters and returns a list of todos.
+//
+// Example:
+//
+//   req: GET /todos/
+//   res: 200 [
+//          {"id": 1, "title": "Learn Go", "completed": false},
+//          {"id": 2, "title": "Buy bread", "completed": true}
+//        ]
+func ListTodos(w http.ResponseWriter, r *http.Request, c context.Context) error {
+	res, err := todo.All(c)
+	if err != nil {
+		log.Errorf(c, "ListTodos: %v", err)
+		return err
+	}
+	return writeJson(w, http.StatusOK, res)
+}
+
+// NewTodo handles POST requests on /todos.
+// The request body must contain a JSON object with a Title field.
+// The status code of the response is used to indicate any error.
+//
+// Examples:
+//
+//   req: POST /todos/ {"title": ""}
+//   res: 400 empty title
+//
+//   req: POST /todos/ {"title": "Buy bread"}
+//   res: 201
+func NewTodo(w http.ResponseWriter, r *http.Request, c context.Context) error {
+	req := struct{ Title string }{}
+	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
+		return badRequest{err}
+	}
+	t, err := todo.NewTodo(req.Title)
+	if err != nil {
+		return badRequest{err}
+	}
+	log.Infof(c, "Saving new todo: %v", t)
```

Comment:
Logging methods that were previously on appengine.Context were moved to the logging package a month or two ago. Still, I get that we don't need it. Deleted.